driver-sync/src/test/functional/com/mongodb/client/AbstractTransactionsTest.java

```diff
@@ -455,4 +486,54 @@ private boolean canRunTests() {
             return false;
         }
     }
+
+    private class TargetedFailPoint {
+        private final BsonDocument failPointDocument;
+        private final MongoDatabase adminDB;
+
+        TargetedFailPoint(final BsonDocument operation) {
+            final BsonDocument arguments = operation.getDocument("arguments", new BsonDocument());
+            final ClientSession clientSession = sessionsMap.get(arguments.getString("session").getValue());
+
+            if (clientSession.getPinnedMongosAddress() != null) {
+                MongoClientSettings.Builder builder = MongoClientSettings.builder()
+                        .applyToClusterSettings(new Block<ClusterSettings.Builder>() {
+                            @Override
+                            public void apply(final ClusterSettings.Builder builder) {
+                                builder.hosts(singletonList(clientSession.getPinnedMongosAddress()));
+                            }
+                        });
+                MongoClient mongoClient = MongoClients.create(MongoClientSettings.builder(builder.build())
+                        .applyToSocketSettings(new Block<SocketSettings.Builder>() {
+                            @Override
+                            public void apply(final SocketSettings.Builder builder) {
+                                builder.readTimeout(5, TimeUnit.SECONDS);
+                            }
+                        })
+                        .build());
+
+                adminDB = mongoClient.getDatabase("admin");
+            } else {
+                adminDB = null;
+            }
+            failPointDocument = arguments.getDocument("failPoint");
+        }
+
+        public void executeFailPoint() {
+            executeCommand(failPointDocument);
+        }
+
+        public void disableFailPoint() {
+            BsonDocument disableDoc = failPointDocument.append("mode", new BsonString("off"));
+            executeCommand(disableDoc);
```

Comment:
This looks slightly different than the code in cleanup() that disables the top-level failpoint.  Can you explain why?

Reply:
I read the structure of the `targetedFailPoint` incorrectly. I have changed the document to emulate what is seen in the cleanup() code.